### PR TITLE
fix ES unpacker when the 2nd OptoRX is disabled, but the 1st and 3rd …

### PR DESCRIPTION
**Pull request by cmkuo**
This is to fix the problem when the second ES OptoRX is disabled, but the first and third OptoRX are enabled in the ES unpacker. The problem would cause the data in third OptoRX are not unpacked.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test
@cmkuo is there a run/file to test this in data?

**Comment by cmkuo** [comment]
@slava77 run 256843 is one of the runs affected by the bug.
If you would like to test on single file, you can try root://eoscms//eos/cms/store/data/Run2015D/DoubleEG/RAW/v1/000/256/843/00000/00E5B2E5-AA5E-E511-B223-02163E0142DD.root

**Comment by slava77** [comment]
... I restarted the tests; let's see if they complete this time

File: EventFilter/ESRawToDigi/src/ESUnpacker.cc [hunk 1]
```diff
@@ -208,16 +208,31 @@ void ESUnpacker::interpretRawData(int fedId, const FEDRawData & rawData, ESRawDa
...
+    enableOptoRX[NenableOptoRX] = 0;
+    NenableOptoRX++;
+  }
+  if (optoRX1_ == 128) {
+    enableOptoRX[NenableOptoRX] = 1;
+    NenableOptoRX++;
+  }
+  if (optoRX2_ == 128) {
+    enableOptoRX[NenableOptoRX] = 2;
+    NenableOptoRX++;
```
Comment (cvuosalo) [inline EventFilter/ESRawToDigi/src/ESUnpacker.cc]: NenableOptoRX is not used after this line, so the increment would seem to serve no purpose. Please delete it.

File: EventFilter/ESRawToDigi/src/ESUnpacker.cc [hunk 2]
```diff
@@ -208,16 +208,31 @@ void ESUnpacker::interpretRawData(int fedId, const FEDRawData & rawData, ESRawDa
...
+  }
 
   // Event data
-  int opto = 0;
+  int iopto = 0;
+  int opto = -1;
   for (const Word64* word=(header+dccWords+1); word!=trailer; ++word) {
     if (debug_) std::cout<<"Event : "<<print(*word)<<std::endl;
```
Comment (cvuosalo) [inline EventFilter/ESRawToDigi/src/ESUnpacker.cc]: Please convert cout to LogDebug(). cout is not thread-safe.

File: EventFilter/ESRawToDigi/src/ESUnpacker.cc [hunk 3]
```diff
@@ -242,10 +257,12 @@ void ESUnpacker::interpretRawData(int fedId, const FEDRawData & rawData, ESRawDa
     } else if (head == 6) {
       optoBC = (*word >> 32) & m16;
       optoEC = (*word >> 48) & m8;      
+
+      opto = enableOptoRX[iopto];
       if (opto==0) ESDCCHeader.setOptoBC0(optoBC);
       else if (opto==1) ESDCCHeader.setOptoBC1(optoBC);
       else if (opto==2) ESDCCHeader.setOptoBC2(optoBC);
-      opto++;
+      iopto++;
```
Comment (cvuosalo) [inline EventFilter/ESRawToDigi/src/ESUnpacker.cc]: I can't tell if iopto can reach 3 or more. Please make sure it doesn't exceed 2, to protect against an out-of-bounds error:
```
if (iopto < 2)
   ++iopto;
```

**Comment by cvuosalo** [comment]
@cmkuo:
I ran prompt reco on the file you suggested with and without your fix. What's an easy way to tell from the output files that your fix worked correctly?

**Comment by slava77** [comment]
On 10/6/15 2:57 PM, Carl Vuosalo wrote:
> @cmkuo <https://github.com/cmkuo>:
> I ran prompt reco on the file you suggested with and without your fix.
> What's an easy way to tell from the output files that your fix worked
> correctly?

I would think that DQM plots will change, probably even the fwlite plots
as well.
Carl, do you see any difference at all?

**Comment by cvuosalo** [comment]
@slava77:
Yes, I found tiny differences in the fwlite plots for ECAL preshower rechit , PF candidate, PF MET, and ak4PFjet quantities. Also, these product size differences show up:

```
-------------------------------------------------------------
 or, B       new, B      delta, B   delta, %   deltaJ, %    branch 
-------------------------------------------------------------
   122054 ->      122781        727      0.6   0.06     recoPFClusters_particleFlowClusterPS__RECO.
    44236 ->       44368        132      0.3   0.01     TrackingRecHitsOwned_generalTracks__RECO.
    86458 ->       86985        528      0.6   0.04     EcalRecHitsSorted_ecalPreshowerRecHit_EcalRecHitsES_RECO.
-------------------------------------------------------------
  1291104 ->     1292600       1496             0.1     ALL BRANCHES
```

It looks like the missing OptoRX data has been restored, so I think the PR is working correctly. After the code fixes are made and new Jenkins tests are passed, Reco should be able to approve this PR.

**Comment by slava77** [comment]
I  expect that some 2D occupancy plots in DQM will be quite clear about
the fix
(localized to the problematic area of ES)

**Comment by cvuosalo** [comment]
DQM plots do show the expected differences. These plots are made in comparison to baseline  CMSSW_7_4_13.

In this occupancy plot, note the right plot has a large white irregular rectangular region on the bottom right, while this region is smaller in the left plot.
occupancy
mismatch

The two following rechit plots show an increase in the number of rechits as the missed OptoRX are restored by this PR.
rechits1
rechits2

**Comment by slava77** [comment]
On 10/7/15 7:58 AM, Carl Vuosalo wrote:
> DQM plots do show the expected differences. These plots are made in
> comparison to baseline CMSSW_7_4_13.
> 
> In this occupancy plot, note the right plot has a large white irregular
> rectangular region on the bottom right, while this region is smaller in
> the left plot.

It looks like the other way around: right hand plot  (usually the "new")
has the white rectangle  smaller.
[probably a typo]

**Comment by cvuosalo** [comment]
@slava77: I fixed the typo (right/left) in my comment on the DQM plots.

**Comment by cmkuo** [comment]
Many thanks for the DQM plots ! These results are as what we expect with the fixes to ES unpacker

**Comment by cvuosalo** [comment]
@cmkuo: I will approve this PR now. Please create a new PR in 76X (or 77X if available) to fix the code problems I noted above.

**Comment by cvuosalo** [comment]
+1

For #11613 5cfff3ebc355e19f0a8b5581c4461be98a9a181d

Fixing an ES unpacker bug where the third OptoRX is missed in certain cases.

The code changes are satisfactory, and Jenkins [tests]() against baseline CMSSW_7_4_X_2015-10-02-1100 show no significant differences. A test of prompt reco using a file known to reproduce the bug shows the expected differences, as shown in previous comments. There are still some minor code issues raised above that should be addressed in a new PR for 76X (or 77X).

**Comment by cmkuo** [comment]
@cvuosalo the new fixes with your comment are committed to git now

**Comment by slava77** [comment]
@cmkuo we will need similar updates in 75X and 76X.
Please update existing PRs (#11615 in 75X) or create a new one (in 76X)
Thank you.

**Comment by cmkuo** [comment]
@slava77 
https://github.com/cms-sw/cmssw/pull/11615 for 75X is updated
Somehow I cannot update 76X one so I create a new PR
https://github.com/cms-sw/cmssw/pull/11693

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/11613	545e2b5
- compared to the previous signoff, the code review comments were addressed. The same updates were applied here https://github.com/cms-sw/cmssw/compare/5cfff3e...545e2b5 as in 75X and in 76X #11693 
- jenkins tests pass and comparisons with the baseline show no differences

@cvuosalo is offline today, I took the liberty to sign off in case there is an early 74X build

**Comment by davidlange6** [comment]
+1

**Pull request merged**